In [66]:
import pandas as pd
import numpy as np
import warnings

warnings.filterwarnings('ignore')

In [67]:
states=pd.read_csv("D:\Data_Science\Veri_Bilimi_Bootcamp\odevler\states_all.csv")
states.columns

Index(['PRIMARY_KEY', 'STATE', 'YEAR', 'ENROLL', 'TOTAL_REVENUE',
       'FEDERAL_REVENUE', 'STATE_REVENUE', 'LOCAL_REVENUE',
       'TOTAL_EXPENDITURE', 'INSTRUCTION_EXPENDITURE',
       'SUPPORT_SERVICES_EXPENDITURE', 'OTHER_EXPENDITURE',
       'CAPITAL_OUTLAY_EXPENDITURE', 'GRADES_PK_G', 'GRADES_KG_G',
       'GRADES_4_G', 'GRADES_8_G', 'GRADES_12_G', 'GRADES_1_8_G',
       'GRADES_9_12_G', 'GRADES_ALL_G', 'AVG_MATH_4_SCORE', 'AVG_MATH_8_SCORE',
       'AVG_READING_4_SCORE', 'AVG_READING_8_SCORE'],
      dtype='object')

# Soru:1 Veri kümesinde yer alan değişkenlerinin tiplerini ve her bir değişkenin eksik (null) değer oranını bulun.

## Cevap 1-1 : Değişken tipleri, object, int ve float

In [68]:
states.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1715 entries, 0 to 1714
Data columns (total 25 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   PRIMARY_KEY                   1715 non-null   object 
 1   STATE                         1715 non-null   object 
 2   YEAR                          1715 non-null   int64  
 3   ENROLL                        1224 non-null   float64
 4   TOTAL_REVENUE                 1275 non-null   float64
 5   FEDERAL_REVENUE               1275 non-null   float64
 6   STATE_REVENUE                 1275 non-null   float64
 7   LOCAL_REVENUE                 1275 non-null   float64
 8   TOTAL_EXPENDITURE             1275 non-null   float64
 9   INSTRUCTION_EXPENDITURE       1275 non-null   float64
 10  SUPPORT_SERVICES_EXPENDITURE  1275 non-null   float64
 11  OTHER_EXPENDITURE             1224 non-null   float64
 12  CAPITAL_OUTLAY_EXPENDITURE    1275 non-null   float64
 13  GRA

.info () metodunun çıktısı verinin 25 sütun ve 1715 satırdan oluştuğunu söylüyor. Sütunlar arasında sadece YEAR                  tamsayı, PRIMARY_KEY ve STATE metin ve diğerleri ondalıklı sayı gibi görünüyor.

In [69]:
states.isnull().head()

PRIMARY_KEY  STATE   YEAR  ENROLL  TOTAL_REVENUE  FEDERAL_REVENUE  \
0        False  False  False    True          False            False   
1        False  False  False    True          False            False   
2        False  False  False    True          False            False   
3        False  False  False    True          False            False   
4        False  False  False    True          False            False   

   STATE_REVENUE  LOCAL_REVENUE  TOTAL_EXPENDITURE  INSTRUCTION_EXPENDITURE  \
0          False          False              False                    False   
1          False          False              False                    False   
2          False          False              False                    False   
3          False          False              False                    False   
4          False          False              False                    False   

   ...  GRADES_4_G  GRADES_8_G  GRADES_12_G  GRADES_1_8_G  GRADES_9_12_G  \
0  ...       False       False        False          True           True   
1  ...       False       False        False          True           True   
2  ...       False       False        False          True           True   
3  ...       False       False        False          True           True   
4  ...       False       False        False          True           True   

   GRADES_ALL_G  AVG_MATH_4_SCORE  AVG_MATH_8_SCORE  AVG_READING_4_SCORE  \
0         False             False             False                False   
1         False              True              True                 True   
2         False             False             False                False   
3         False             False             False                False   
4         False             False             False                False   

   AVG_READING_8_SCORE  
0                 True  
1                 True  
2                 True  
3                 True  
4                 True  

[5 rows x 25 columns]

In [27]:
for sutun_adi in states.columns:
    print("{} sütunundaki benzersiz değerler : {}".format(sutun_adi, states[sutun_adi].unique()))

PRIMARY_KEY sütunundaki benzersiz değerler : ['1992_ALABAMA' '1992_ALASKA' '1992_ARIZONA' ... '2019_WEST_VIRGINIA'
 '2019_WISCONSIN' '2019_WYOMING']
STATE sütunundaki benzersiz değerler : ['ALABAMA' 'ALASKA' 'ARIZONA' 'ARKANSAS' 'CALIFORNIA' 'COLORADO'
 'CONNECTICUT' 'DELAWARE' 'DISTRICT_OF_COLUMBIA' 'FLORIDA' 'GEORGIA'
 'HAWAII' 'IDAHO' 'ILLINOIS' 'INDIANA' 'IOWA' 'KANSAS' 'KENTUCKY'
 'LOUISIANA' 'MAINE' 'MARYLAND' 'MASSACHUSETTS' 'MICHIGAN' 'MINNESOTA'
 'MISSISSIPPI' 'MISSOURI' 'MONTANA' 'NEBRASKA' 'NEVADA' 'NEW_HAMPSHIRE'
 'NEW_JERSEY' 'NEW_MEXICO' 'NEW_YORK' 'NORTH_CAROLINA' 'NORTH_DAKOTA'
 'OHIO' 'OKLAHOMA' 'OREGON' 'PENNSYLVANIA' 'RHODE_ISLAND' 'SOUTH_CAROLINA'
 'SOUTH_DAKOTA' 'TENNESSEE' 'TEXAS' 'UTAH' 'VERMONT' 'VIRGINIA'
 'WASHINGTON' 'WEST_VIRGINIA' 'WISCONSIN' 'WYOMING' 'DODEA' 'NATIONAL']
YEAR sütunundaki benzersiz değerler : [1992 1993 1994 1995 1996 1997 1998 1999 2000 2001 2002 2003 2004 2005
 2006 2007 2008 2009 2010 2011 2012 2013 2014 2015 2016 1986 1987 1988
 1989 19

Sadece YEAR tamsayı, PRIMARY_KEY ve STATE metin ve diğerleri ondalıklı sayı olduğunu gördük. Hatalı değerler ney her biri için ona bakalım.

In [28]:
for sutun_adi in ["PRIMARY_KEY", "STATE"]:
    print("{} sütunu için problemli değerler : ".format(sutun_adi))
    hatali_degerler = []
    for deger in states[sutun_adi]:
        try:
            str(deger)
        except:
            hatali_degerler.append(deger)
    print(set(hatali_degerler))

PRIMARY_KEY sütunu için problemli değerler : 
set()
STATE sütunu için problemli değerler : 
set()


Hatalı değer çıkmadı.

In [29]:
for sutun_adi in ["YEAR"]:
    print("{} sütunu için problemli değerler : ".format(sutun_adi))
    hatali_degerler = []
    for deger in states[sutun_adi]:
        try:
            int(deger)
        except:
            hatali_degerler.append(deger)
    print(set(hatali_degerler))

YEAR sütunu için problemli değerler : 
set()


Hatalı değer çıkmadı.

In [30]:
for sutun_adi in states.loc[:,"ENROLL":"AVG_READING_8_SCORE"]:
    print("{} sütunu için problemli değerler : ".format(sutun_adi))
    hatali_degerler = []
    for deger in states[sutun_adi]:
        try:
            float(deger)
        except:
            hatali_degerler.append(deger)
    print(set(hatali_degerler))

ENROLL sütunu için problemli değerler : 
set()
TOTAL_REVENUE sütunu için problemli değerler : 
set()
FEDERAL_REVENUE sütunu için problemli değerler : 
set()
STATE_REVENUE sütunu için problemli değerler : 
set()
LOCAL_REVENUE sütunu için problemli değerler : 
set()
TOTAL_EXPENDITURE sütunu için problemli değerler : 
set()
INSTRUCTION_EXPENDITURE sütunu için problemli değerler : 
set()
SUPPORT_SERVICES_EXPENDITURE sütunu için problemli değerler : 
set()
OTHER_EXPENDITURE sütunu için problemli değerler : 
set()
CAPITAL_OUTLAY_EXPENDITURE sütunu için problemli değerler : 
set()
GRADES_PK_G sütunu için problemli değerler : 
set()
GRADES_KG_G sütunu için problemli değerler : 
set()
GRADES_4_G sütunu için problemli değerler : 
set()
GRADES_8_G sütunu için problemli değerler : 
set()
GRADES_12_G sütunu için problemli değerler : 
set()
GRADES_1_8_G sütunu için problemli değerler : 
set()
GRADES_9_12_G sütunu için problemli değerler : 
set()
GRADES_ALL_G sütunu için problemli değerler : 
set()
A

## Cevap 1-2: Her bir değişkenin eksik (null) değer oranı aşağıda yüzde olarak bulunmaktadır.

In [31]:
states.isnull().sum()*100/states.shape[0]

PRIMARY_KEY                      0.000000
STATE                            0.000000
YEAR                             0.000000
ENROLL                          28.629738
TOTAL_REVENUE                   25.655977
FEDERAL_REVENUE                 25.655977
STATE_REVENUE                   25.655977
LOCAL_REVENUE                   25.655977
TOTAL_EXPENDITURE               25.655977
INSTRUCTION_EXPENDITURE         25.655977
SUPPORT_SERVICES_EXPENDITURE    25.655977
OTHER_EXPENDITURE               28.629738
CAPITAL_OUTLAY_EXPENDITURE      25.655977
GRADES_PK_G                     10.087464
GRADES_KG_G                      4.839650
GRADES_4_G                       4.839650
GRADES_8_G                       4.839650
GRADES_12_G                      4.839650
GRADES_1_8_G                    40.524781
GRADES_9_12_G                   37.551020
GRADES_ALL_G                     4.839650
AVG_MATH_4_SCORE                67.055394
AVG_MATH_8_SCORE                64.897959
AVG_READING_4_SCORE             62

# Soru2: Verimizde yıl (year) sütunu olduğunu farketmişsinizdir. Şimdilik yıl verisini unutun ve her bir gözlemin aynı yıl içerisinde yapıldığını farz edin. Her bir değişken için eksik değerleri nasıl doldurabileceğinizi düşünün. Eksik değerleri bir değerle doldurmak hangi değişkenler için anlamlı, hangileri için anlamsızdır?

## Cevap 2: Float tipi değerler için anlamlı, "Primary Key" ve "State" için boşlukları doldurmak anlamlı olmayabilir.

# Soru3: Şimdi zaman faktörünü dikkate alma zamanı! 2. sorudaki cevabınızı tekrar gözden geçirin ve eksik verileri o yıl içerisinde gözlemlenen değerlere dayanarak doldurun. Örneğin, bir değeri ortalama değer ile doldurmak isterseniz, o yılın ortalamasını hesaplayın.

## Cevap3: Ne kadar eksik değer olduğunu yüzde olarak aşağıda görüyoruz.

In [70]:
education_df = states.copy()
#education_df

In [71]:
education_df.isnull().sum()*100/education_df.shape[0] 

PRIMARY_KEY                      0.000000
STATE                            0.000000
YEAR                             0.000000
ENROLL                          28.629738
TOTAL_REVENUE                   25.655977
FEDERAL_REVENUE                 25.655977
STATE_REVENUE                   25.655977
LOCAL_REVENUE                   25.655977
TOTAL_EXPENDITURE               25.655977
INSTRUCTION_EXPENDITURE         25.655977
SUPPORT_SERVICES_EXPENDITURE    25.655977
OTHER_EXPENDITURE               28.629738
CAPITAL_OUTLAY_EXPENDITURE      25.655977
GRADES_PK_G                     10.087464
GRADES_KG_G                      4.839650
GRADES_4_G                       4.839650
GRADES_8_G                       4.839650
GRADES_12_G                      4.839650
GRADES_1_8_G                    40.524781
GRADES_9_12_G                   37.551020
GRADES_ALL_G                     4.839650
AVG_MATH_4_SCORE                67.055394
AVG_MATH_8_SCORE                64.897959
AVG_READING_4_SCORE             62

In [34]:
means = education_df.groupby("YEAR").ENROLL.mean()
means

YEAR
1986              NaN
1987              NaN
1988              NaN
1989              NaN
1990              NaN
1991              NaN
1992              NaN
1993    802054.333333
1994    801383.960784
1995    858973.039216
1996    873930.607843
1997    888379.745098
1998    897509.058824
1999    905101.470588
2000    910446.705882
2001    916296.274510
2002    923818.196078
2003    933396.862745
2004    938327.117647
2005    942259.960784
2006    947078.980392
2007    949109.254902
2008    949047.392157
2009    946249.313725
2010    946396.254902
2011    946587.745098
2012    945379.392157
2013    947012.176471
2014    948607.529412
2015    951263.450980
2016    952388.764706
2017              NaN
2019              NaN
Name: ENROLL, dtype: float64

In [60]:
for sutun_adi in fill_list:
    education2_df.loc[:, col].fillna(education2_df.loc[:, col].mean(), inplace=True)
education2_df 

PRIMARY_KEY          STATE  YEAR         ENROLL  TOTAL_REVENUE  \
0           1992_ALABAMA        ALABAMA  1992  917541.566176   2.678885e+06   
1            1992_ALASKA         ALASKA  1992  917541.566176   1.049591e+06   
2           1992_ARIZONA        ARIZONA  1992  917541.566176   3.258079e+06   
3          1992_ARKANSAS       ARKANSAS  1992  917541.566176   1.711959e+06   
4        1992_CALIFORNIA     CALIFORNIA  1992  917541.566176   2.626002e+07   
...                  ...            ...   ...            ...            ...   
1710       2019_VIRGINIA       VIRGINIA  2019  917541.566176   9.102045e+06   
1711     2019_WASHINGTON     WASHINGTON  2019  917541.566176   9.102045e+06   
1712  2019_WEST_VIRGINIA  WEST_VIRGINIA  2019  917541.566176   9.102045e+06   
1713      2019_WISCONSIN      WISCONSIN  2019  917541.566176   9.102045e+06   
1714        2019_WYOMING        WYOMING  2019  917541.566176   9.102045e+06   

      FEDERAL_REVENUE  STATE_REVENUE  LOCAL_REVENUE  TOTAL_EXPENDITURE  \
0        3.041770e+05   1.659028e+06   7.156800e+05       2.653798e+06   
1        1.067800e+05   7.207110e+05   2.221000e+05       9.724880e+05   
2        2.978880e+05   1.369815e+06   1.590376e+06       3.401580e+06   
3        1.785710e+05   9.587850e+05   5.746030e+05       1.743022e+06   
4        2.072470e+06   1.654651e+07   7.641041e+06       2.713883e+07   
...               ...            ...            ...                ...   
1710     7.677799e+05   4.223743e+06   4.110522e+06       9.206242e+06   
1711     7.677799e+05   4.223743e+06   4.110522e+06       9.206242e+06   
1712     7.677799e+05   4.223743e+06   4.110522e+06       9.206242e+06   
1713     7.677799e+05   4.223743e+06   4.110522e+06       9.206242e+06   
1714     7.677799e+05   4.223743e+06   4.110522e+06       9.206242e+06   

      INSTRUCTION_EXPENDITURE  ...     GRADES_4_G     GRADES_8_G  \
0                1.481703e+06  ...   57948.000000   58025.000000   
1                4.983620e+05  ...    9748.000000    8789.000000   
2                1.435908e+06  ...   55433.000000   49081.000000   
3                9.643230e+05  ...   34632.000000   36011.000000   
4                1.435892e+07  ...  418418.000000  363296.000000   
...                       ...  ...            ...            ...   
1710             4.768010e+06  ...   69166.507966   68106.537377   
1711             4.768010e+06  ...   69166.507966   68106.537377   
1712             4.768010e+06  ...   69166.507966   68106.537377   
1713             4.768010e+06  ...   69166.507966   68106.537377   
1714             4.768010e+06  ...   69166.507966   68106.537377   

        GRADES_12_G   GRADES_1_8_G  GRADES_9_12_G  GRADES_ALL_G  \
0      41167.000000  577954.717647  282069.084967  7.316340e+05   
1       6714.000000  577954.717647  282069.084967  1.224870e+05   
2      37410.000000  577954.717647  282069.084967  6.734770e+05   
3      27651.000000  577954.717647  282069.084967  4.414900e+05   
4     270675.000000  577954.717647  282069.084967  5.254844e+06   
...             ...            ...            ...           ...   
1710   58420.870711  577954.717647  282069.084967  9.139695e+05   
1711   58420.870711  577954.717647  282069.084967  9.139695e+05   
1712   58420.870711  577954.717647  282069.084967  9.139695e+05   
1713   58420.870711  577954.717647  282069.084967  9.139695e+05   
1714   58420.870711  577954.717647  282069.084967  9.139695e+05   

      AVG_MATH_4_SCORE  AVG_MATH_8_SCORE  AVG_READING_4_SCORE  \
0                208.0             252.0                207.0   
1                  NaN               NaN                  NaN   
2                215.0             265.0                209.0   
3                210.0             256.0                211.0   
4                208.0             261.0                202.0   
...                ...               ...                  ...   
1710             247.0             287.0                224.0   
1711             240.0           

In [ ]:
means = education_df.groupby("YEAR").ENROLL.mean()
means

In [39]:
fill_list = ["ENROLL", "TOTAL_REVENUE", "FEDERAL_REVENUE", 
             "STATE_REVENUE", "LOCAL_REVENUE", "TOTAL_EXPENDITURE", 
             "INSTRUCTION_EXPENDITURE", "SUPPORT_SERVICES_EXPENDITURE",
             "OTHER_EXPENDITURE", "CAPITAL_OUTLAY_EXPENDITURE", "GRADES_PK_G", 
             "GRADES_KG_G", "GRADES_4_G", "GRADES_8_G", "GRADES_12_G", "GRADES_1_8_G", 
             "GRADES_9_12_G", "GRADES_ALL_G"]

In [37]:
years = education_df["YEAR"].unique()

## aşağıda NA olan veriler için içinde bulunduğu sütunların ortalaması verilmiştir.

In [40]:
for col in fill_list:
    education_df.loc[:, col].fillna(education_df.loc[:, col].mean(), inplace=True)

education_df

PRIMARY_KEY          STATE  YEAR         ENROLL  TOTAL_REVENUE  \
0           1992_ALABAMA        ALABAMA  1992  917541.566176   2.678885e+06   
1            1992_ALASKA         ALASKA  1992  917541.566176   1.049591e+06   
2           1992_ARIZONA        ARIZONA  1992  917541.566176   3.258079e+06   
3          1992_ARKANSAS       ARKANSAS  1992  917541.566176   1.711959e+06   
4        1992_CALIFORNIA     CALIFORNIA  1992  917541.566176   2.626002e+07   
...                  ...            ...   ...            ...            ...   
1710       2019_VIRGINIA       VIRGINIA  2019  917541.566176   9.102045e+06   
1711     2019_WASHINGTON     WASHINGTON  2019  917541.566176   9.102045e+06   
1712  2019_WEST_VIRGINIA  WEST_VIRGINIA  2019  917541.566176   9.102045e+06   
1713      2019_WISCONSIN      WISCONSIN  2019  917541.566176   9.102045e+06   
1714        2019_WYOMING        WYOMING  2019  917541.566176   9.102045e+06   

      FEDERAL_REVENUE  STATE_REVENUE  LOCAL_REVENUE  TOTAL_EXPENDITURE  \
0        3.041770e+05   1.659028e+06   7.156800e+05       2.653798e+06   
1        1.067800e+05   7.207110e+05   2.221000e+05       9.724880e+05   
2        2.978880e+05   1.369815e+06   1.590376e+06       3.401580e+06   
3        1.785710e+05   9.587850e+05   5.746030e+05       1.743022e+06   
4        2.072470e+06   1.654651e+07   7.641041e+06       2.713883e+07   
...               ...            ...            ...                ...   
1710     7.677799e+05   4.223743e+06   4.110522e+06       9.206242e+06   
1711     7.677799e+05   4.223743e+06   4.110522e+06       9.206242e+06   
1712     7.677799e+05   4.223743e+06   4.110522e+06       9.206242e+06   
1713     7.677799e+05   4.223743e+06   4.110522e+06       9.206242e+06   
1714     7.677799e+05   4.223743e+06   4.110522e+06       9.206242e+06   

      INSTRUCTION_EXPENDITURE  ...     GRADES_4_G     GRADES_8_G  \
0                1.481703e+06  ...   57948.000000   58025.000000   
1                4.983620e+05  ...    9748.000000    8789.000000   
2                1.435908e+06  ...   55433.000000   49081.000000   
3                9.643230e+05  ...   34632.000000   36011.000000   
4                1.435892e+07  ...  418418.000000  363296.000000   
...                       ...  ...            ...            ...   
1710             4.768010e+06  ...   69166.507966   68106.537377   
1711             4.768010e+06  ...   69166.507966   68106.537377   
1712             4.768010e+06  ...   69166.507966   68106.537377   
1713             4.768010e+06  ...   69166.507966   68106.537377   
1714             4.768010e+06  ...   69166.507966   68106.537377   

        GRADES_12_G   GRADES_1_8_G  GRADES_9_12_G  GRADES_ALL_G  \
0      41167.000000  577954.717647  282069.084967  7.316340e+05   
1       6714.000000  577954.717647  282069.084967  1.224870e+05   
2      37410.000000  577954.717647  282069.084967  6.734770e+05   
3      27651.000000  577954.717647  282069.084967  4.414900e+05   
4     270675.000000  577954.717647  282069.084967  5.254844e+06   
...             ...            ...            ...           ...   
1710   58420.870711  577954.717647  282069.084967  9.139695e+05   
1711   58420.870711  577954.717647  282069.084967  9.139695e+05   
1712   58420.870711  577954.717647  282069.084967  9.139695e+05   
1713   58420.870711  577954.717647  282069.084967  9.139695e+05   
1714   58420.870711  577954.717647  282069.084967  9.139695e+05   

      AVG_MATH_4_SCORE  AVG_MATH_8_SCORE  AVG_READING_4_SCORE  \
0                208.0             252.0                207.0   
1                  NaN               NaN                  NaN   
2                215.0             265.0                209.0   
3                210.0             256.0                211.0   
4                208.0             261.0                202.0   
...                ...               ...                  ...   
1710             247.0             287.0                224.0   
1711             240.0           

## aşağıda ise NA olan veriler için içinde bulunduğu sütunların yıllara göre ortalaması verilmiştir. 

In [55]:
education2_df = education_df.copy()
years = education2_df["YEAR"].unique()

for col in fill_list:
    for year in years:
        YIL_SART = education2_df["YEAR"] == year
        education2_df.loc[YIL_SART, col] = education2_df.loc[YIL_SART, col].fillna(education2_df[YIL_SART][col].mean())

education2_df

PRIMARY_KEY          STATE  YEAR         ENROLL  TOTAL_REVENUE  \
0           1992_ALABAMA        ALABAMA  1992  917541.566176   2.678885e+06   
1            1992_ALASKA         ALASKA  1992  917541.566176   1.049591e+06   
2           1992_ARIZONA        ARIZONA  1992  917541.566176   3.258079e+06   
3          1992_ARKANSAS       ARKANSAS  1992  917541.566176   1.711959e+06   
4        1992_CALIFORNIA     CALIFORNIA  1992  917541.566176   2.626002e+07   
...                  ...            ...   ...            ...            ...   
1710       2019_VIRGINIA       VIRGINIA  2019  917541.566176   9.102045e+06   
1711     2019_WASHINGTON     WASHINGTON  2019  917541.566176   9.102045e+06   
1712  2019_WEST_VIRGINIA  WEST_VIRGINIA  2019  917541.566176   9.102045e+06   
1713      2019_WISCONSIN      WISCONSIN  2019  917541.566176   9.102045e+06   
1714        2019_WYOMING        WYOMING  2019  917541.566176   9.102045e+06   

      FEDERAL_REVENUE  STATE_REVENUE  LOCAL_REVENUE  TOTAL_EXPENDITURE  \
0        3.041770e+05   1.659028e+06   7.156800e+05       2.653798e+06   
1        1.067800e+05   7.207110e+05   2.221000e+05       9.724880e+05   
2        2.978880e+05   1.369815e+06   1.590376e+06       3.401580e+06   
3        1.785710e+05   9.587850e+05   5.746030e+05       1.743022e+06   
4        2.072470e+06   1.654651e+07   7.641041e+06       2.713883e+07   
...               ...            ...            ...                ...   
1710     7.677799e+05   4.223743e+06   4.110522e+06       9.206242e+06   
1711     7.677799e+05   4.223743e+06   4.110522e+06       9.206242e+06   
1712     7.677799e+05   4.223743e+06   4.110522e+06       9.206242e+06   
1713     7.677799e+05   4.223743e+06   4.110522e+06       9.206242e+06   
1714     7.677799e+05   4.223743e+06   4.110522e+06       9.206242e+06   

      INSTRUCTION_EXPENDITURE  ...     GRADES_4_G     GRADES_8_G  \
0                1.481703e+06  ...   57948.000000   58025.000000   
1                4.983620e+05  ...    9748.000000    8789.000000   
2                1.435908e+06  ...   55433.000000   49081.000000   
3                9.643230e+05  ...   34632.000000   36011.000000   
4                1.435892e+07  ...  418418.000000  363296.000000   
...                       ...  ...            ...            ...   
1710             4.768010e+06  ...   69166.507966   68106.537377   
1711             4.768010e+06  ...   69166.507966   68106.537377   
1712             4.768010e+06  ...   69166.507966   68106.537377   
1713             4.768010e+06  ...   69166.507966   68106.537377   
1714             4.768010e+06  ...   69166.507966   68106.537377   

        GRADES_12_G   GRADES_1_8_G  GRADES_9_12_G  GRADES_ALL_G  \
0      41167.000000  577954.717647  282069.084967  7.316340e+05   
1       6714.000000  577954.717647  282069.084967  1.224870e+05   
2      37410.000000  577954.717647  282069.084967  6.734770e+05   
3      27651.000000  577954.717647  282069.084967  4.414900e+05   
4     270675.000000  577954.717647  282069.084967  5.254844e+06   
...             ...            ...            ...           ...   
1710   58420.870711  577954.717647  282069.084967  9.139695e+05   
1711   58420.870711  577954.717647  282069.084967  9.139695e+05   
1712   58420.870711  577954.717647  282069.084967  9.139695e+05   
1713   58420.870711  577954.717647  282069.084967  9.139695e+05   
1714   58420.870711  577954.717647  282069.084967  9.139695e+05   

      AVG_MATH_4_SCORE  AVG_MATH_8_SCORE  AVG_READING_4_SCORE  \
0                208.0             252.0                207.0   
1                  NaN               NaN                  NaN   
2                215.0             265.0                209.0   
3                210.0             256.0                211.0   
4                208.0             261.0                202.0   
...                ...               ...                  ...   
1710             247.0             287.0                224.0   
1711             240.0           

In [56]:
means = education2_df.groupby("YEAR")["ENROLL"].mean()
means

YEAR
1986    917541.566176
1987    917541.566176
1988    917541.566176
1989    917541.566176
1990    917541.566176
1991    917541.566176
1992    917541.566176
1993    802054.333333
1994    805767.266648
1995    858973.039216
1996    875576.304384
1997    888379.745098
1998    898265.002497
1999    905101.470588
2000    910714.436459
2001    916296.274510
2002    923581.342120
2003    932798.549667
2004    938327.117647
2005    941327.191176
2006    947078.980392
2007    947918.021365
2008    949047.392157
2009    945166.002497
2010    946396.254902
2011    945491.662875
2012    945379.392157
2013    945900.077969
2014    948607.529412
2015    949990.927026
2016    952388.764706
2017    917541.566176
2019    917541.566176
Name: ENROLL, dtype: float64

## aşağıda görüldüğü gibi NA olan verileri sayıları 0a inmiştir

In [57]:
education2_df.isnull().sum()*100/education2_df.shape[0]  

PRIMARY_KEY                      0.000000
STATE                            0.000000
YEAR                             0.000000
ENROLL                           0.000000
TOTAL_REVENUE                    0.000000
FEDERAL_REVENUE                  0.000000
STATE_REVENUE                    0.000000
LOCAL_REVENUE                    0.000000
TOTAL_EXPENDITURE                0.000000
INSTRUCTION_EXPENDITURE          0.000000
SUPPORT_SERVICES_EXPENDITURE     0.000000
OTHER_EXPENDITURE                0.000000
CAPITAL_OUTLAY_EXPENDITURE       0.000000
GRADES_PK_G                      0.000000
GRADES_KG_G                      0.000000
GRADES_4_G                       0.000000
GRADES_8_G                       0.000000
GRADES_12_G                      0.000000
GRADES_1_8_G                     0.000000
GRADES_9_12_G                    0.000000
GRADES_ALL_G                     0.000000
AVG_MATH_4_SCORE                67.055394
AVG_MATH_8_SCORE                64.897959
AVG_READING_4_SCORE             62

In [83]:
education2_df.ENROLL.isnull().sum()

0

In [61]:
education2_df.ENROLL.describe()

count    1.715000e+03
mean     9.175416e+05
std      9.008956e+05
min      4.386600e+04
25%      4.646880e+05
50%      9.175416e+05
75%      9.175416e+05
max      6.307022e+06
Name: ENROLL, dtype: float64

# Eksik değerleri enterpolasyon yaparak dolduralım

In [74]:
education3_df = education_df.copy()

In [76]:
education3_df = education3_df.interpolate()
education3_df

PRIMARY_KEY          STATE  YEAR   ENROLL  TOTAL_REVENUE  \
0           1992_ALABAMA        ALABAMA  1992      NaN      2678885.0   
1            1992_ALASKA         ALASKA  1992      NaN      1049591.0   
2           1992_ARIZONA        ARIZONA  1992      NaN      3258079.0   
3          1992_ARKANSAS       ARKANSAS  1992      NaN      1711959.0   
4        1992_CALIFORNIA     CALIFORNIA  1992      NaN     26260025.0   
...                  ...            ...   ...      ...            ...   
1710       2019_VIRGINIA       VIRGINIA  2019  94511.0      2044669.0   
1711     2019_WASHINGTON     WASHINGTON  2019  94511.0      2044669.0   
1712  2019_WEST_VIRGINIA  WEST_VIRGINIA  2019  94511.0      2044669.0   
1713      2019_WISCONSIN      WISCONSIN  2019  94511.0      2044669.0   
1714        2019_WYOMING        WYOMING  2019  94511.0      2044669.0   

      FEDERAL_REVENUE  STATE_REVENUE  LOCAL_REVENUE  TOTAL_EXPENDITURE  \
0            304177.0      1659028.0       715680.0          2653798.0   
1            106780.0       720711.0       222100.0           972488.0   
2            297888.0      1369815.0      1590376.0          3401580.0   
3            178571.0       958785.0       574603.0          1743022.0   
4           2072470.0     16546514.0      7641041.0         27138832.0   
...               ...            ...            ...                ...   
1710         123012.0      1175899.0       745758.0          2034229.0   
1711         123012.0      1175899.0       745758.0          2034229.0   
1712         123012.0      1175899.0       745758.0          2034229.0   
1713         123012.0      1175899.0       745758.0          2034229.0   
1714         123012.0      1175899.0       745758.0          2034229.0   

      INSTRUCTION_EXPENDITURE  ...  GRADES_4_G  GRADES_8_G  GRADES_12_G  \
0                   1481703.0  ...     57948.0     58025.0      41167.0   
1                    498362.0  ...      9748.0      8789.0       6714.0   
2                   1435908.0  ...     55433.0     49081.0      37410.0   
3                    964323.0  ...     34632.0     36011.0      27651.0   
4                  14358922.0  ...    418418.0    363296.0     270675.0   
...                       ...  ...         ...         ...          ...   
1710                 921494.0  ...      7537.0      7086.0       6405.0   
1711                 921494.0  ...      7537.0      7086.0       6405.0   
1712                 921494.0  ...      7537.0      7086.0       6405.0   
1713                 921494.0  ...      7537.0      7086.0       6405.0   
1714                 921494.0  ...      7537.0      7086.0       6405.0   

      GRADES_1_8_G  GRADES_9_12_G  GRADES_ALL_G  AVG_MATH_4_SCORE  \
0              NaN            NaN      731634.0             208.0   
1              NaN            NaN      122487.0             211.5   
2              NaN            NaN      673477.0             215.0   
3              NaN            NaN      441490.0             210.0   
4              NaN            NaN     5254844.0             208.0   
...            ...            ...           ...               ...   
1710       58998.0        27361.0       94258.0             247.0   
1711       58998.0        27361.0       94258.0             240.0   
1712       58998.0        27361.0       94258.0             231.0   
1713       58998.0        27361.0       94258.0             242.0   
1714       58998.0        27361.0       94258.0             246.0   

      AVG_MATH_8_SCORE  AVG_READING_4_SCORE  AVG_READING_8_SCORE  
0                252.0                207.0                  NaN  
1                258.5                208.0                  NaN  
2                265.0                209.0                  NaN  
3                256.0                211.0                  NaN  
4                261.0                202.0                  NaN  
...                ...                  ...                  ...  
1710             287.0                224.0     

# interpolasyon sonucunda yine bazı değerlerin eksik olduğu görünmektedir

In [78]:
education3_df.isnull().sum()*100/education3_df.shape[0] 

PRIMARY_KEY                      0.000000
STATE                            0.000000
YEAR                             0.000000
ENROLL                           2.973761
TOTAL_REVENUE                    0.000000
FEDERAL_REVENUE                  0.000000
STATE_REVENUE                    0.000000
LOCAL_REVENUE                    0.000000
TOTAL_EXPENDITURE                0.000000
INSTRUCTION_EXPENDITURE          0.000000
SUPPORT_SERVICES_EXPENDITURE     0.000000
OTHER_EXPENDITURE                2.973761
CAPITAL_OUTLAY_EXPENDITURE       0.000000
GRADES_PK_G                      0.000000
GRADES_KG_G                      0.000000
GRADES_4_G                       0.000000
GRADES_8_G                       0.000000
GRADES_12_G                      0.000000
GRADES_1_8_G                    14.868805
GRADES_9_12_G                   14.868805
GRADES_ALL_G                     0.000000
AVG_MATH_4_SCORE                 0.000000
AVG_MATH_8_SCORE                 0.000000
AVG_READING_4_SCORE              0

In [82]:
education3_df.ENROLL.isnull().sum()

51

In [77]:
education3_df.ENROLL.describe()

count    1.664000e+03
mean     6.999133e+05
std      9.840401e+05
min      4.386600e+04
25%      9.451100e+04
50%      3.295345e+05
75%      8.768110e+05
max      6.307022e+06
Name: ENROLL, dtype: float64

# yılların kendi içindeki ortalama değerini kullanmak, interpolasyondan daha iyi sonuç vermektedir